Implement inline-math with an arbitrary number of operators

alt-romes opened the pull request:
Known bug: When mixing 0x with 0b with decimals in multiple-operators math sometimes the history is not well displayed

Can someone test this out before I merge it into master?

Thank you
~romes

DevManu-de commented:
Hi,

I have tested it and experience the same bug as you mentioned but in addition to that I also noticed some wrong calculations when mixing the three bases.
Example: `0x13+131-0b1010` outputs `Decimal: 19`
 `History: 0x13 + 0b1010 - 0b1010`
> Known bug: When mixing 0x with 0b with decimals in multiple-operators math sometimes the history is not well displayed

I didnt take a deep look into the code but I advise not to merge it but its your choice.

DevManu-de

alt-romes commented:
You're right, it seems we need to do more work here.

There's also something not mentioned before: There are no priorities. 1 + 2 * 3 will do 1+2 followed by * 3.
This is what I expect of it, but I don't want to mislead anyone.

Tomorrow I'll try to fix the inline with multiple number representations

For now we won't merge

alt-romes commented:
Hello,

I'm pretty sure I've fixed this.
If anyone could test it and detect errors before I merge it with master it would be good.

I believe it's working really well - and I've also cleaned up the code.

Thank you,
~romes

DevManu-de commented:
Hi @alt-romes,

I have tested the pr mainly for memory leaks but the rest seems to be working really good. :)

**IMPORTANT: This is just the part of the pr. There are some more memory leaks but they were fixed in #20.**

But we definitely have to test it again after the pr is merged.

Note: The only input given was `1+1+1`
```
==181263== Memcheck, a memory error detector
==181263== Copyright (C) 2002-2017, and GNU GPL'd, by Julian Seward et al.
==181263== Using Valgrind-3.16.1 and LibVEX; rerun with -h for copyright info
==181263== Command: ./pcalc
==181263==
==181263==
==181263== HEAP SUMMARY:
==181263==     in use at exit: 1,390,280 bytes in 281 blocks
==181263==   total heap usage: 294 allocs, 13 frees, 1,400,019 bytes allocated
==181263==
==181263== 3 bytes in 1 blocks are definitely lost in loss record 2 of 56
==181263==    at 0x483A77F: malloc (vg_replace_malloc.c:307)
==181263==    by 0x496FB4E: strdup (in /usr/lib/libc-2.32.so)
==181263==    by 0x10A13C: process_input (main.c:236)
==181263==    by 0x109FBC: main (main.c:150)
==181263==
==181263== 5 bytes in 1 blocks are possibly lost in loss record 5 of 56
==181263==    at 0x483A77F: malloc (vg_replace_malloc.c:307)
==181263==    by 0x496FB4E: strdup (in /usr/lib/libc-2.32.so)
==181263==    by 0x10A13C: process_input (main.c:236)
==181263==    by 0x109FBC: main (main.c:150)
==181263==
==181263== 5 bytes in 1 blocks are definitely lost in loss record 6 of 56
==181263==    at 0x483A77F: malloc (vg_replace_malloc.c:307)
==181263==    by 0x496FB4E: strdup (in /usr/lib/libc-2.32.so)
==181263==    by 0x10A02B: process_input (main.c:187)
==181263==    by 0x109FBC: main (main.c:150)
==181263==
==181263== LEAK SUMMARY:
==181263==    definitely lost: 168 bytes in 3 blocks
==181263==    indirectly lost: 0 bytes in 0 blocks
==181263==      possibly lost: 5 bytes in 1 blocks
==181263==    still reachable: 1,390,107 bytes in 277 blocks
==181263==         suppressed: 0 bytes in 0 blocks
==181263== Reachable blocks (those to which a pointer was found) are not shown.
==181263== To see them, rerun with: --leak-check=full --show-leak-kinds=all
==181263==
==181263== ERROR SUMMARY: 4 errors from 4 contexts (suppressed: 0 from 0)
```

Maybe you can taka a look at it.

DevManu-de

DevManu-de commented on src/main.c:
```diff
@@ -184,47 +177,27 @@ void process_input(numberstack* numbers, operation** current_op, char* in) {
+        /* Before the strtok replaces the operator with \0 when searching for tokens
+         * Save the string starting at the operator so we can reuse after handling the first number
+         */
+        char * in_saved = strdup(op);
```
in_saved is never `free()`

alt-romes replied:
I've fixed this @DevManu-de. Do you think you could check it one more time?
I believe it's ready to merge.
Afterwards you could retry the memory leak tests with the merged version as well.

DevManu-de commented on src/main.c:
```diff
@@ -237,18 +210,93 @@ void process_input(numberstack* numbers, operation** current_op, char* in) {
+            /* Before the strtok replaces the operator with \0 when searching for tokens
+             * Save the string starting from right after the operator so we can search it later for more ops
+             */
+            in_saved = strdup(op) + 1;
```
in_saved shold be `free()` before giving it a new pointer address.

DevManu-de reviewed (changes requested):
Without actually checking, just an assumption.
These problems mentioned should fix the memory leaks mentioned in my previous comment.

alt-romes commented:
I'm going to merge it - the tests are all okay.

We still need to check the final master branch for leaks
